n3fit/src/evolven3fit/utils.py

```diff
@@ -0,0 +1,173 @@
+from scipy.interpolate import interp1d
+import numpy as np
+import math
+from reportengine.compat import yaml
+import shutil
+from validphys.pdfbases import PIDS_DICT
+import pathlib
+
+
+class LhapdfLike:
+    """
+    Class which emulates lhapdf but only for an initial condition PDF (i.e. with only one q2 value)
+    """
+
+    def __init__(self, pdf_grid, q20, x_grid):
+        self.pdf_grid = pdf_grid
+        self.q20 = q20
+        self.x_grid = x_grid
+        self.funcs = [
+            interp1d(self.x_grid, self.pdf_grid[pid], kind="cubic")
+            for pid in range(len(PIDS_DICT))
+        ]
+
+    def xfxQ2(self, pid, x, q2):
+        """Return the value of the PDF for the requested pid and x value. If the requested q2
+        value is different from the (only) value available, it raises an error.
+
+        Parameters
+        ----------
+
+            pid: int
+                pid index of particle
+            x: float
+                x-value
+            q2: float
+                Q square value
+
+        Returns
+        -------
+            : float
+            x * PDF value
+        """
+        if not math.isclose(q2, self.q20, rel_tol=1e-6):
+            raise ValueError("The q2 requested is not the fitting scale of this pdf")
+        return self.funcs[list(PIDS_DICT.values()).index(PIDS_DICT[pid])](x)
+
+    def hasFlavor(self, pid):
+        """Check if the requested pid is in the PDF."""
+        return pid in PIDS_DICT.keys()
+
+
+def read_runcard(usr_path):
+    """Read the runcard and return the relevant information for evolven3fit"""
+    return yaml.safe_load((usr_path / "filter.yml").read_text())
+
+
+def get_theoryID_from_runcard(usr_path):
+    """Return the theoryID from the runcard"""
+    # read the runcard
+    my_runcard = read_runcard(usr_path)
+    return my_runcard["theory"]["theoryid"]
+
+
+def generate_q2grid(Q0, Qfin, Q_points, match_dict):
+    """Generate the q2grid used in the final evolved pdfs or use the default grid if Qfin or Q_points is
+    not provided.
+    """
+    if Qfin is None:
+        if Q_points is None:
+            return (
+                np.array(
+                    [
+                        1.6500000e00,
+                        1.7874388e00,
+                        1.9429053e00,
+                        2.1193749e00,
+                        2.3204100e00,
+                        2.5502944e00,
+                        2.8142025e00,
+                        3.1184122e00,
+                        3.4705775e00,
+                        3.8800751e00,
+                        4.3584516e00,
+                        4.9200000e00,
+                        4.9200000e00,
```

Comment:
As per the discussion during the code meeting @andreab1997 @felixhekhorn at the quark thresholds it is important to tell eko where to match from (and we do want the two subgrids).

Reply:
So you are saying that this grid is correct right? As you can see I have also changed the way in which the grid is produced when the user provides the final point and the number of points. Now, we actually produce one subgrid for each matching needed (each of them is logarithmically scaled inside its range) and the matching point is both the final of its subgrid and the initial of the next one.

Reply:
The grid yes. The way the eko is computed I don't know (from what you say... I would guess yes?)

Reply:
I believe so but I am not sure. Maybe it is worth asking to @AleCandido and @felixhekhorn

Reply:
The question was about the subgrids inside LHAPDF (called blocks inside banana): the subgrids correspond to a given FNS - by default one 4-flavor and one 5-flavor, with the bottom mass separating the two. We need to make sure that the last point in the 4-flavor patch is computed with 4 flavors and the first point in 5-flavor with 5 flavors. EKO can explicitly distinguish the two - or you have to make sure that the point do lay a epsilon from the matching point

Reply:
Yeah, it was a bit subtle: when we tried to address with epsilon, we got poor performances because of infinitesimal evolution (blowing up numerical errors/integration convergence). Thus, there are options in the EKO runcard that you can use to specify the number of flavors (but only for the first patch, i.e. with the option `nf0`).